sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala

```diff
@@ -589,6 +587,50 @@ object Decimal {
     }
   }
 
+  private def stringToJavaBigDecimal(str: UTF8String): (JavaBigDecimal, Int) = {
+    // According the benchmark test,  `s.toString.trim` is much faster than `s.trim.toString`.
+    // Please refer to https://github.com/apache/spark/pull/26640
+    val bigDecimal = new JavaBigDecimal(str.toString.trim)
+    val precision = if (bigDecimal.scale < 0) {
+      bigDecimal.precision - bigDecimal.scale
+    } else {
+      bigDecimal.precision
+    }
+    (bigDecimal, precision)
```

Comment:
This is a performance-critical code path. Can we avoid repeatedly creating Tuple2 objects? We can add a private method to calculate precision.